src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -1011,25 +1010,41 @@ private void BlockDownloaded(uint256 blockHash, Block block, int peerId)
                     throw new InvalidOperationException("Unsolicited block");
                 }
 
-                if (block != null)
+                chainedHeader = this.chainedHeaderTree.GetChainedHeader(blockHash);
+                if (chainedHeader == null)
                 {
-                    chainedHeader = this.chainedHeaderTree.GetChainedHeader(blockHash);
+                    lock (this.blockRequestedLock)
+                    {
+                        this.callbacksByBlocksRequestedHash.Remove(blockHash);
+                    }
+
+                    this.logger.LogTrace("(-)[CHAINED_HEADER_NOT_FOUND]");
+                    return;
+                }
 
-                    if (chainedHeader == null)
+                if (block == null)
+                {
+                    // A race conditions exists where if we attempted a download of a block but all the peers disconnected and then a peer presented the header
+                    // again, we dont re-download the block.
+                    if (chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockRequired)
                     {
+                        // We need to remove the current callback so that it can be re-assigned for download.
                         lock (this.blockRequestedLock)
                         {
                             this.callbacksByBlocksRequestedHash.Remove(blockHash);
                         }
 
-                        this.logger.LogTrace("(-)[CHAINED_HEADER_NOT_FOUND]");
-                        return;
+                        reassignDownload = true;
                     }
+                    else
+                        this.logger.LogTrace("(-)[CHAINEDHEADER_INVALID_BLOCK_AVAILABILITY_STATE]:{0}", chainedHeader.BlockDataAvailability);
                 }
-                else
-                {
-                    this.logger.LogDebug("Block '{0}' failed to be delivered.", blockHash);
-                }
+            }
+
+            if (reassignDownload)
+            {
+                this.logger.LogWarning("Downloading block for '{0}' failed, it will be enqueued again.", chainedHeader);
+                this.DownloadBlocks(new[] { chainedHeader }, this.ProcessDownloadedBlock);
```

Comment:
here you are adding a callback. then on 1081 what was added will be removed. Original race condition is preserved. 
returning after the reassigning should fix the problem

Reply:
Ah yes OK, we should just return. here. Will fix.